### DISCUSSION TITLE
@TmsLink configuration

### DISCUSSION THREAD

**Discussion by ghost** [allure-framework/allure2#2076] [Ideas & Feedback]
I assume that many, like me, implement a large number of autotests based on test cases from a TMS.
In this case, it is convenient to name the test methods in accordance with the test case ID, for example

 ```
    @TmsLink("T1")
    @DisplayName("Some test name")
    @Test
    public void T1() {
    }
```

As you can see, we need to add an annotation that will duplicate the code.

Can we make it possible for @TmsLink to implicitly use test method names?

**Comment by baev** [comment]
You can implement such an extension using `io.qameta.allure.listener.TestLifecycleListener`:

1. create an implementation of `TestLifecycleListener`, eg `org.example.AutoTmsLinkTestLifecycleListener`
1. implement `beforeTestStop` method:
    ```java
     @Override
    public void beforeTestStop(final TestResult result) {
        final boolean containsLinkAlready = result.getLinks()
                .stream().anyMatch(l -> ResultsUtils.TMS_LINK_TYPE.equals(l.getType()));

        if (!containsLinkAlready) {
            final Optional<String> maybeMethodName = result.getLabels().stream()
                    .filter(l -> ResultsUtils.TEST_METHOD_LABEL_NAME.equals(l.getName()))
                    .map(Label::getValue)
                    .findFirst();

            maybeMethodName.ifPresent(methodName -> {
                // change it the way you want
                if (methodName.matches("[A-Z]+\\d+")) {
                    result.getLinks().add(ResultsUtils.createTmsLink(methodName));
                }
            });
        }
    }
   ```
1.  create `META-INF/services/io.qameta.allure.listener.TestLifecycleListener` file in your test resources folder and specify the reference to `AutoTmsLinkTestLifecycleListener` (SPI)
